### PR TITLE
#26: Zoltan2: MatrixAdapter refactor API

### LINKED ISSUE #26: Zoltan2:  MatrixAdapter refactor API
TpetraRowMatrixAdapter
- [x] : Update user API (kokkos version) (RowIDs, CRSView, RowWeightsView, ColumnIDsView, CCSView, ColumnWeightsView, IDsView, WeightsView)
- [x]  : Provide the Unit Test 
- [x] : Documentation

XpetraCrsMatrixAdapter
- [x] : Update user API (kokkos version) (RowIDs, CRSView, RowWeightsView, ColumnIDsView, CCSView, ColumnWeightsView, IDsView, WeightsView)
- [ ]  : Provide the Unit Test 
- [x] : Documentation

### PR DESCRIPTION
**Pull request by cwschilly**
Fixes #26 and adds TpetraCrsMatrixAdapter

### CODE REVIEW

**Comment by cwschilly** [comment]
This PR is almost ready for review; I just want to run it using the `graph.cpp` example to make sure everything works as expected (this example requires use of TPL's that I haven't installed yet)

**Comment by cwschilly** [comment]
Since the `doMigration()` method was originally defined in `XpetraTraits.hpp`, and there is no corresponding `TpetraTraits` file, it had to be redefined within the new `TpetraCrsMatrixAdapter.hpp`. It is also defined in `TpetraRowMatrixAdapter.hpp`, but I considered this duplication to be the best option to minimize changes to external API.

File: packages/zoltan2/core/src/input/Zoltan2_TpetraCrsMatrixAdapter.hpp [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+  }
+  Z2_FORWARD_EXCEPTIONS;
+
+  // Move the rows, creating a new matrix.
+  out = this->doMigration(in, numNewRows, importList.getRawPtr());
+}
+
+}  //namespace Zoltan2
+
+#endif
```
Comment (stmcgovern) [inline packages/zoltan2/core/src/input/Zoltan2_TpetraCrsMatrixAdapter.hpp]: This file (and a couple others) drops the trailing new line. Let's not change that, please.

File: packages/zoltan2/core/src/input/Zoltan2_XpetraCrsMatrixAdapter.hpp [hunk 2]
```diff
@@ -400,4 +403,4 @@ template <typename User, typename UserCoord>
 
 }  //namespace Zoltan2
 
-#endif
+#endif
```
Comment (stmcgovern) [inline packages/zoltan2/core/src/input/Zoltan2_XpetraCrsMatrixAdapter.hpp]: lost new line again.

File: packages/zoltan2/test/core/helpers/Zoltan2_TestHelpers.hpp [hunk 3]
```diff
@@ -198,4 +198,4 @@ void TestDeviceHostView(const DeviceType &deviceView,
 #include <PrintData.hpp>
 #include <UserInputForTests.hpp>
 
-#endif
+#endif
```
Comment (stmcgovern) [inline packages/zoltan2/test/core/helpers/Zoltan2_TestHelpers.hpp]: lost new line again

File: packages/zoltan2/test/core/unit/input/TpetraCrsMatrixInput.cpp [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+    verifyInputAdapter(*newInput, *newM);
+
+  } catch (std::exception &e) {
+    std::cout << e.what() << std::endl;
+    return EXIT_FAILURE;
+  }
+
+  PrintFromRoot("PASS");
+
+  }
```
Comment (stmcgovern) [inline packages/zoltan2/test/core/unit/input/TpetraCrsMatrixInput.cpp]: please add new line

File: packages/zoltan2/test/core/unit/input/TpetraRowMatrixInput.cpp [hunk 5]
```diff
@@ -98,169 +105,205 @@ void printMatrix(RCP<const Comm<int> > &comm, zlno_t nrows,
...
   }
 
-  /////////////////////////////////////////////////////////////
-  // DONE
+  PrintFromRoot("PASS");
 
-  if (rank==0)
-    std::cout << "PASS" << std::endl;
-}
+  }
```
Comment (stmcgovern) [inline packages/zoltan2/test/core/unit/input/TpetraRowMatrixInput.cpp]: new line needed

**Review by stmcgovern — CHANGES_REQUESTED** [review]
looks good - please fix the new lines issue

**Comment by JacobDomagala** [comment]
I believe the changes will break NGA-develop branch. Please rebase with recent changes to NGA-develop and fix compilation errors

**Comment by JacobDomagala** [comment]
See https://github.com/NexGenAnalytics/Trilinos/commit/e066b2f8200d2ba533595516e302f5ec70b34754 for reference on how to deal with Weights getters (we want to support both 2D and 1D versions - https://github.com/NexGenAnalytics/Trilinos/issues/14#issuecomment-1651878217)

File: packages/zoltan2/core/src/input/Zoltan2_TpetraCrsMatrixAdapter.hpp (hunk 1 above)
Comment (cwschilly) [reply packages/zoltan2/core/src/input/Zoltan2_TpetraCrsMatrixAdapter.hpp]: done

File: packages/zoltan2/core/src/input/Zoltan2_XpetraCrsMatrixAdapter.hpp (hunk 2 above)
Comment (cwschilly) [reply packages/zoltan2/core/src/input/Zoltan2_XpetraCrsMatrixAdapter.hpp]: done

File: packages/zoltan2/test/core/unit/input/TpetraCrsMatrixInput.cpp (hunk 4 above)
Comment (cwschilly) [reply packages/zoltan2/test/core/unit/input/TpetraCrsMatrixInput.cpp]: done

File: packages/zoltan2/test/core/unit/input/TpetraRowMatrixInput.cpp (hunk 5 above)
Comment (cwschilly) [reply packages/zoltan2/test/core/unit/input/TpetraRowMatrixInput.cpp]: done

**Review by stmcgovern — APPROVED** [review]
It passes local tests with the new API. Thanks @cwschilly!

**Pull request merged**